File: backend/src/models/calificacion.model.ts
import { pool } from '../config/database';

export interface Calificacion {
  id_calificacion: string;
  id_sesion: string;
  id_tutor: string;
  id_tutorado: string;
  puntuacion: number;
  comentario: string | null;
  fecha: Date;
}

export interface NuevaCalificacionInput {
  id_sesion: string;
  id_tutor: string;
  id_tutorado: string;
  puntuacion: number;
  comentario?: string;
}

export async function crearCalificacion(datos: NuevaCalificacionInput): Promise<Calificacion> {
  const { id_sesion, id_tutor, id_tutorado, puntuacion, comentario } = datos;

  const query = `
    INSERT INTO calificacion (id_sesion, id_tutor, id_tutorado, puntuacion, comentario)
    VALUES ($1, $2, $3, $4, $5)
    RETURNING *;
  `;

  const result = await pool.query(query, [
    id_sesion,
    id_tutor,
    id_tutorado,
    puntuacion,
    comentario || null,
  ]);
  return result.rows[0];
}

// HU-27: listado de calificaciones recibidas por un tutor, con datos del tutorado y la sesión.
export interface CalificacionConDatos extends Calificacion {
  nombre_tutorado: string;
  tema_sesion: string;
}

export async function listarCalificacionesPorTutor(id_tutor: string): Promise<CalificacionConDatos[]> {
  const query = `
    SELECT c.*, u.nombre_completo AS nombre_tutorado, s.tema AS tema_sesion
    FROM calificacion c
    JOIN usuario u ON u.id_usuario = c.id_tutorado
    JOIN sesion_estudio s ON s.id_sesion = c.id_sesion
    WHERE c.id_tutor = $1
    ORDER BY c.fecha DESC;
  `;
  const result = await pool.query(query, [id_tutor]);
  return result.rows;
}

File: app/lib/features/calificaciones/data/models/calificacion_model.dart
class CalificacionModel {
  final String idCalificacion;
  final String idSesion;
  final String idTutor;
  final String idTutorado;
  final int puntuacion;
  final String? comentario;
  final DateTime fecha;
  final String? nombreTutorado;
  final String? temaSesion;

  CalificacionModel({
    required this.idCalificacion,
    required this.idSesion,
    required this.idTutor,
    required this.idTutorado,
    required this.puntuacion,
    this.comentario,
    required this.fecha,
    this.nombreTutorado,
    this.temaSesion,
  });

  factory CalificacionModel.fromJson(Map<String, dynamic> json) {
    return CalificacionModel(
      idCalificacion: json['id_calificacion'] as String,
      idSesion: json['id_sesion'] as String,
      idTutor: json['id_tutor'] as String,
      idTutorado: json['id_tutorado'] as String,
      puntuacion: json['puntuacion'] as int,
      comentario: json['comentario'] as String?,
      fecha: DateTime.parse(json['fecha'] as String),
      nombreTutorado: json['nombre_tutorado'] as String?,
      temaSesion: json['tema_sesion'] as String?,
    );
  }
}

class ReputacionModel {
  final String idUsuario;
  final String nombreCompleto;
  final double reputacion;
  final int totalCalificaciones;
  final List<CalificacionModel> calificaciones;

  ReputacionModel({
    required this.idUsuario,
    required this.nombreCompleto,
    required this.reputacion,
    required this.totalCalificaciones,
    required this.calificaciones,
  });

  factory ReputacionModel.fromJson(Map<String, dynamic> json) {
    return ReputacionModel(
      idUsuario: json['id_usuario'] as String,
      nombreCompleto: json['nombre_completo'] as String,
      reputacion: double.parse(json['reputacion'].toString()),
      totalCalificaciones: json['total_calificaciones'] as int,
      calificaciones: (json['calificaciones'] as List)
          .map((e) => CalificacionModel.fromJson(e))
          .toList(),
    );
  }
}
